pup-0008.md

```diff
@@ -0,0 +1,75 @@
+---
+title: Python Code Style and Tools
+author: David Davis
+created: 8-Jun-2019
+status:
+version: 1
+---
+
+## Summary
+
+PUP-8 proposes the adoption of an official Python code style and set of tools for the Pulp
+community.
+
+### Background
+
+While most Pulp projects adhere to PEP 8[0], there is still room for different code styles to emerge
+and different toolsets to be used to enforce style for Python code. This PUP seeks to come to a
+common agreement about which style and tools we use. Specifically, this PUP proposes adoption of
+black[1] along with pydocstyle[2] as the official toolset for formatting and linting Python code.
+
+## Motivation
+
+The Pulp ecosystem contains numerous different coding styles that make it difficult to switch from
+one repo or plugin to another. Moreover, style comments create a lot of noise in PRs and can lead to
+bikeshedding[3]. In order to free developers from having to worry about code style, PUP-8 seeks to
+adopt a single style along with common tooling to aid in writing Python code.
+
+Another added benefit is that using an autoformatter can make it easier for newer Pulp users
+(especially those unfamiliar with Python) to create mergeable PRs. The PR process can be a barrier
+to entry for some users and removing style concerns can help to get PRs merged faster.
+
+## Detailed Design
+
+In order to implement this proposal, black and pydocstyle will be used/enforced in the core
+repositories (pulpcore, pulpcore-plugin, etc). They will also be added to the plugin_template so
+plugins can leverage them. If approved, this PUP can also serve as documentation for what the
+official style and set of linting tools are for Python code in the Pulp community.
+
+## Drawbacks
+
+There are a few drawbacks to using black. The main one is that it is an opinionated style and tool.
+Not everyone is going to agree with all of its style choices. It is worth noting though that other
+popular projects such as django have adopted black[4]. So its style will likely become more
+commonplace in the Python community.
+
+Another concern will be that reformatting all of Pulp's python code will likely touch a lot of
+places within our codebase and create a lot of noise in git history. However, this would be problem
```

Comment:
@daviddavis, is it worth adding that there is basically no way back? (not black specific, it would be the same for any autoformatter). It might be another reason to wait for broader adoption in the community and for black  to go stable first.

Reply:
What do you mean by "no way back"? We can always drop black later if another tool comes along or we decide we don't want to use it.

Reply:
How do you drop it if every commit will be formatted in black way? We can stop using it at some point (which probably wouldn't make much sense if we don't use another autoformatter), however I don't see a way to undo the changes introduced by black.

Reply:
If we adopt another style, we can move from black to that style. Sure, moving back to not having a style is hard but I am not sure I see value in doing that. I will add your concern though to the PUP.